core/src/main/java/io/grpc/internal/BinaryLog.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright 2018, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.internal;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.grpc.CallOptions;
+import io.grpc.Channel;
+import io.grpc.ClientCall;
+import io.grpc.ClientInterceptor;
+import io.grpc.ClientInterceptors;
+import io.grpc.InternalClientInterceptors;
+import io.grpc.InternalServerInterceptors;
+import io.grpc.MethodDescriptor;
+import io.grpc.MethodDescriptor.Marshaller;
+import io.grpc.ServerCallHandler;
+import io.grpc.ServerInterceptor;
+import io.grpc.ServerMethodDefinition;
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+public abstract class BinaryLog implements Closeable {
+  private static final Logger logger = Logger.getLogger(BinaryLog.class.getName());
+  @VisibleForTesting
+  static final Marshaller<InputStream> IDENTITY_MARSHALLER = new IdentityMarshaller();
+
+  private final ClientInterceptor binaryLogShim = new BinaryLogShim();
+
+  /**
+   * Returns a {@link ServerInterceptor} for binary logging. gRPC is free to cache the interceptor,
+   * so the interceptor must be reusable across calls. At runtime, the request and response
+   * marshallers are always {@code Marshaller<InputStream>}.
+   * Returns {@code null} if this method is not binary logged.
+   */
+  // TODO(zpencer): ensure the interceptor properly handles retries and hedging
+  @Nullable
+  public abstract ServerInterceptor getServerInterceptor(String fullMethodName);
```

Comment:
Can this just take the MethodDescriptor as the input?

Reply:
This may reveal too much about the method, because the binary log design spec allows users to configure it using per gRPC method name options. Also, there is some magic happening that swaps out the Marshaller to be InputStream based, so the method descriptor can not match the service definition and the actual interceptor's Marshaller at the same time.

Reply:
The MD has a minor advantage from an API point of view, because the method name is unambiguous..   It also provides the opportunity for caching (like we do for the `:path:` header).  Lastly it allows additional info to be provided.  For example, we could add a MD attribute called "loggable" which controls if certain methods should be logged.  I'm not saying we will do these things, but it leaves the door open for it.

I won't keep pushing on this if you want to keep it string, but please consider it.

Reply:
I think making the parameter a `MethodDescriptor<?, ?>` should make it obvious that its marshallers should not be used.